### PR TITLE
feat: public cairo_call precompile

### PR DESCRIPTION
**Pull request by enitrat**
A public cairo_call precompile at address 0x75004, with delegate/callcode disabled.

reworked a bit the solidity contracts that are used solely for testing purposed.

todo next:
- rework the public cairo lib to use 75003 and 75004 (multicall, singlecall cairo precompiles) instead of 75001
- recompile the pragma contracts with these new precomps

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1509)

### CODE REVIEW

File: cairo_zero/kakarot/precompiles/precompiles_helpers.cairo [hunk 1]
```diff
@@ -101,8 +109,9 @@ namespace PrecompilesHelpers {
         let range_check_ptr = [ap - 2];
         let authorized = [ap - 1];
 
-        // Ensure that calls to CAIRO_BATCH_CALL_PRECOMPILE are not made through a delegatecall / callcode.
-        if (precompile_address == Constants.CAIRO_BATCH_CALL_PRECOMPILE) {
+        // Ensure that calls to CAIRO_CALL_PRECOMPILE are not made through a delegatecall / callcode.
```
Comment (obatirou) [inline cairo_zero/kakarot/precompiles/precompiles_helpers.cairo:112]:
```suggestion
        // Ensure that calls to CAIRO_CALL_PRECOMPILE or CAIRO_CALL_PRECOMPILE  are not made through a delegatecall / callcode.
```

File: cairo_zero/kakarot/precompiles/precompiles_helpers.cairo [hunk 2]
```diff
@@ -9,8 +9,8 @@ from kakarot.storages import Kakarot_authorized_cairo_precompiles_callers
 const LAST_ETHEREUM_PRECOMPILE_ADDRESS = 0x0a;
 const FIRST_ROLLUP_PRECOMPILE_ADDRESS = Constants.P256VERIFY_PRECOMPILE;
 const LAST_ROLLUP_PRECOMPILE_ADDRESS = Constants.P256VERIFY_PRECOMPILE;
-const FIRST_KAKAROT_PRECOMPILE_ADDRESS = Constants.CAIRO_CALL_PRECOMPILE;
-const LAST_KAKAROT_PRECOMPILE_ADDRESS = Constants.CAIRO_BATCH_CALL_PRECOMPILE;
+const FIRST_KAKAROT_PRECOMPILE_ADDRESS = Constants.CAIRO_WHITELISTED_CALL_PRECOMPILE;
+const LAST_KAKAROT_PRECOMPILE_ADDRESS = Constants.CAIRO_CALL_PRECOMPILE;
```
Comment (obatirou) [inline cairo_zero/kakarot/precompiles/precompiles_helpers.cairo:13]: Might need to update the ones in python in `constants.py`

File: cairo_zero/kakarot/precompiles/kakarot_precompiles.cairo [hunk 3]
```diff
@@ -32,13 +32,14 @@ const CAIRO_MESSAGE_GAS = 5000;
 
 namespace KakarotPrecompiles {
     // @notice Executes a cairo contract/class.
-    // @dev Requires a whitelisted caller, as this could be called by CALLCODE / DELEGATECALL
+    // @dev If called with 0x75001, the caller _must_ be whitelisted, as this could be called by CALLCODE / DELEGATECALL.
+    // @dev It called with 0x75004, the caller can be anyone, as DELEGATECALL / CALLCODE are not allowed.
```
Comment (ClementWalter) [inline cairo_zero/kakarot/precompiles/kakarot_precompiles.cairo:36]:
```suggestion
    // @dev If called with 0x75004, the caller can be anyone, as DELEGATECALL / CALLCODE are not allowed.
```

File: solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    /// be the caller of this function.
+    function delegateCallIncrementCairoCounter() external {
+        uint256[] memory data = new uint256[](0);
+        cairoCounter.delegatecallCairo("inc", data);
+    }
+
+    /// @notice Calls the Cairo contract to increment its internal counter
+    function incrementCairoCounter() external {
+        uint256[] memory data = new uint256[](0);
+        cairoCounter.callCairo("inc", data);
```
Comment (ClementWalter) [inline solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol:33]:
```suggestion
        cairoCounter.callCairo("inc");
```

File: solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        data[0] = uint256(newCounterLow);
+        data[1] = uint256(newCounterHigh);
+        cairoCounter.callCairo("set_counter", data);
+    }
+
+    /// @notice Calls the Cairo contract to get the (starknet) address of the last caller
+    /// @return lastCaller The starknet address of the last caller
+    function getLastCaller() external view returns (uint256 lastCaller) {
+        uint256[] memory data = new uint256[](0);
+        bytes memory returnData = cairoCounter.staticcallCairo("get_last_caller", data);
```
Comment (ClementWalter) [inline solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol:55]:
```suggestion
        bytes memory returnData = cairoCounter.staticcallCairo("get_last_caller");
```

File: solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        return abi.decode(returnData, (uint256));
+    }
+
+    /// @notice Calls the Cairo contract to increment its internal counter
+    /// @dev The delegatecall preserves the caller's context, so the caller's address will
+    /// be the caller of this function.
+    /// @dev Should always fail, as MulticallCairo does not support delegatecalls.
+    function incrementCairoCounterDelegatecall() external {
+        uint256[] memory data = new uint256[](0);
+        cairoCounter.delegatecallCairo("inc", data);
```
Comment (ClementWalter) [inline solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol:66]:
```suggestion
        cairoCounter.delegatecallCairo("inc");
```

Comment (ClementWalter) [reply solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol:66]: same as `delegateCallIncrementCairoCounter` ,

File: solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /// @notice Calls the Cairo contract to increment its internal counter
+    /// @dev Called with a regular call, the caller's address will be this contract's address
+    /// @dev Should always fail, as MulticallCairo does not support callcode.
+    function incrementCairoCounterCallcode() external {
+        uint256[] memory data = new uint256[](0);
+        cairoCounter.callcodeCairo("inc", data);
```
Comment (ClementWalter) [inline solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol:74]:
```suggestion
        cairoCounter.callcodeCairo("inc");
```

File: solidity_contracts/src/CairoPrecompiles/MulticallCairoPrecompileTest.sol [hunk 8]
```diff
@@ -12,16 +12,14 @@ contract MulticallCairoCounterCaller {
...
-
     constructor(uint256 cairoContractAddress) {
         cairoCounter = cairoContractAddress;
     }
 
     /// @notice Calls the Cairo contract to increment its internal counter
     function incrementCairoCounter() external {
-        cairoCounter.callCairo("inc");
+        uint256[] memory data = new uint256[](0);
+        cairoCounter.callCairo("inc", data);
```
Comment (ClementWalter) [inline solidity_contracts/src/CairoPrecompiles/MulticallCairoPrecompileTest.sol:22]:
```suggestion
        cairoCounter.callCairo("inc");
```

File: solidity_contracts/src/CairoPrecompiles/MulticallCairoPrecompileTest.sol [hunk 9]
```diff
@@ -42,14 +40,16 @@ contract MulticallCairoCounterCaller {
     /// be the caller of this function.
     /// @dev Should always fail, as MulticallCairo does not support delegatecalls.
     function incrementCairoCounterDelegatecall() external {
-        cairoCounter.delegatecallCairo("inc");
+        uint256[] memory data = new uint256[](0);
+        cairoCounter.delegatecallCairo("inc", data);
```
Comment (ClementWalter) [inline solidity_contracts/src/CairoPrecompiles/MulticallCairoPrecompileTest.sol:44]:
```suggestion
        cairoCounter.delegatecallCairo("inc");
```

File: solidity_contracts/src/CairoPrecompiles/MulticallCairoPrecompileTest.sol [hunk 10]
```diff
@@ -42,14 +40,16 @@ contract MulticallCairoCounterCaller {
...
     }
 
     /// @notice Calls the Cairo contract to increment its internal counter
     /// @dev Called with a regular call, the caller's address will be this contract's address
     /// @dev Should always fail, as MulticallCairo does not support callcode.
     function incrementCairoCounterCallcode() external {
-        cairoCounter.callcodeCairo("inc");
+        uint256[] memory data = new uint256[](0);
+        cairoCounter.callcodeCairo("inc", data);
```
Comment (ClementWalter) [inline solidity_contracts/src/CairoPrecompiles/MulticallCairoPrecompileTest.sol:52]:
```suggestion
        cairoCounter.callcodeCairo("inc");
```

File: solidity_contracts/src/CairoPrecompiles/SubContextPrecompile.sol [hunk 11]
```diff
@@ -21,11 +21,11 @@ contract SubContextPrecompile {
 }
 
 contract RevertingSubContext {
-    CairoCounterCaller immutable cairo_counter_caller;
+    WhitelistedCallCairoPrecompileTest immutable cairo_counter_caller;
```
Comment (ClementWalter) [inline solidity_contracts/src/CairoPrecompiles/SubContextPrecompile.sol:24]:
```suggestion
    WhitelistedCallCairoPrecompileTest immutable cairoCounterCaller;
```

File: tests/end_to_end/CairoPrecompiles/test_call_cairo_precompile.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+
+@pytest_asyncio.fixture(scope="module")
+async def cairo_counter(max_fee, deployer):
+    cairo_counter = get_contract("Counter", provider=deployer)
+
+    yield cairo_counter
+
+    tx_hash = await invoke("Counter", "set_counter", 0)
+    await wait_for_transaction(tx_hash)
```
Comment (ClementWalter) [inline tests/end_to_end/CairoPrecompiles/test_call_cairo_precompile.py:19]: `invoke` already waits
```suggestion
```

File: tests/end_to_end/CairoPrecompiles/test_call_cairo_precompile.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+
+@pytest_asyncio.fixture(scope="module")
+async def cairo_counter_caller(cairo_counter):
+    caller_contract = await deploy(
+        "CairoPrecompiles",
+        "CallCairoPrecompileTest",
+        cairo_counter.address,
+    )
+    return caller_contract
```
Comment (ClementWalter) [inline tests/end_to_end/CairoPrecompiles/test_call_cairo_precompile.py:29]:
```suggestion
    return await deploy(
        "CairoPrecompiles",
        "CallCairoPrecompileTest",
        cairo_counter.address,
    )
```

File: tests/end_to_end/CairoPrecompiles/test_call_cairo_precompile.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+@pytest.mark.CairoPrecompiles
+class TestCairoPrecompiles:
+    class TestCounterPrecompiles:
+        async def test_should_increase_counter(self, cairo_counter, owner):
+            prev_count = (await cairo_counter.functions["get"].call()).count
+
+            call = cairo_counter.functions["inc"].prepare_call()
+            tx_data = encode(
+                ["uint256", "uint256", "uint256[]"],
+                [int(cairo_counter.address), int(call.selector), call.calldata],
```
Comment (ClementWalter) [inline tests/end_to_end/CairoPrecompiles/test_call_cairo_precompile.py:42]:
```suggestion
                [int(call.to_addr), int(call.selector), call.calldata],
```

File: solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol (hunk 6 above)
Comment (enitrat) [reply solidity_contracts/src/CairoPrecompiles/CallCairoPrecompileTest.sol:66]: removed the other one

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**